src/server.h

```diff
@@ -2605,10 +2605,10 @@ void listTypeDelRange(robj *o, long start, long stop);
 void unblockClientWaitingData(client *c);
 void popGenericCommand(client *c, int where);
 void listElementsRemoved(client *c, robj *key, int where, robj *o, long count, int signal, int *deleted);
-void listTypeConvertToListpack(robj *o);
 typedef void (*beforeConvertCB)(void *data);
-void listTypeTryConversionForValues(robj *o, robj **argv, int start, int end, beforeConvertCB fn, void *data);
-void listTypeTryConversion(robj *o, int enc, beforeConvertCB fn, void *data);
+void listTypeTryConvertQuicklist(robj *o, int shrinking, beforeConvertCB fn, void *data);
+void listTypeTryConversionForGrowing(robj *o, robj **argv, int start, int end, beforeConvertCB fn, void *data);
+void listTypeTryConversionForShrinking(robj *o, beforeConvertCB fn, void *data);
```

Comment:
i don't like the fact we have 3 interfaces, and not just 2 (one for either direction), or two (one with the argv argument and one without)..
i also don't like the mention of "Quicklist" in the interface (not clear if it's the source or the target).

btw, i'm not sure if my "shrinking" / "growing" suggestion was what you implemented, i meant to only consider going from LP to QL when adding elements, and only consider going from QL to LP when removing elements. i didn't consider using different thresholds, but maybe i don't understand the problem well enough.

anyway, if we keep the current approach, and assuming i understand it correctly (it took me very long):
i understand we actually have a trinary.
1. when we know we added an item and we only consider going LP->QL. that's listTypeTryConversionForGrowing (this one also has an extra set of arguments so you can call it before the item was added)
2. when we know we removed an item and we only consider going QL->LP. that's listTypeTryConversionForShrinking (internally it calls listTypeTryConvertQuicklist with shrinking=1)
3. when we don't know what happened (e.g. item was replaced) and we wanna let the system decide what to do. in this case we call listTypeTryConvertQuicklist directly with shrinking=1.

as i said, it took me quite a bit to realize that, so i suggest changing it.
one option is to only have one interface, that takes a trinary `tip` (-1, 0, 1), or an enum (growing, shrinking, unknown)
besides that we can have a fork of that api without the argv if we want.
i'd rather not mention "quicklist" or "listpack" in the interface.

does that sound ok?

Reply:
> btw, i'm not sure if my "shrinking" / "growing" suggestion was what you implemented, i meant to only consider going from LP to QL when adding elements, and only consider going from QL to LP when removing elements. i didn't consider using different thresholds, but maybe i don't understand the problem well enough.

I guess that you misunderstand `int threshold = shrinking ? 2 : 1;`, it's not using a different threshold, just to reuse the conversion code so that the threshold can be ignored when RDB loading.

Reply:
@oranagra Yes, these interfaces have troubled me for a long time.
If there is only one interface, it may take 7 parameters. I don't like so many parameters.
But first let me try the way you said.

Reply:
@oranagra Only one interface was retained, please see the commit [`6f1b697` (#11303)](https://github.com/redis/redis/pull/11303/commits/6f1b69710b4471f9441ed99f5211f311f0d770b1)

Reply:
I feel that `unknown` doesn't seem relevant, what about `on-demand`?

Reply:
On-demand describes *when* something happens (when needed, when required), not how or why.

Maybe `auto`?

Or if it is about growing and shringing, maybe `both`? or a bitwise-or `GROWING | SHRINKING`...

Reply:
by UNKNOWN, i mean that the caller doesn't know if we're shrinking or growing.
but AUTO is ok with me too (as long as the comment is good enough to mention when to use it and what it does)

Reply:
I like `auto` more.